<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Reputils vignette • Reputils</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Reputils vignette">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Reputils</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="../articles/Prerequisites.html">
    <span class="fa fa-play-circle"></span>
     
    Prerequisites
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Mapping_QC.html">Read alignment QC</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-th-list"></span>
     
    Reference
  </a>
</li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-news"></span>
     
    News
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://bitbucket.org/tanaylab/Reputils/src/default/">
    <span class="fa fa-bitbucket-square"></span>
     
    Bitbucket
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Reputils vignette</h1>
                        <h4 class="author">David Brocks</h4>
            
            <h4 class="date">2019-07-03</h4>
      
      
      <div class="hidden name"><code>Mapping_QC.Rmd</code></div>

    </div>

    
    
<div id="benchmark-read-alignments-to-transposable-elements" class="section level1">
<h1 class="hasAnchor">
<a href="#benchmark-read-alignments-to-transposable-elements" class="anchor"></a>Benchmark read alignments to transposable elements</h1>
<p>In this tutorial, we extract the genomic reference sequence of our read alignments, introduce random base substitutations, and investigate the impact of error frequencies on read mis-mapping.</p>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h2>
<p>We start the workflow by loading Reputils and other libraries, the human hg38 <em>BSgenome</em> object and TE intervals into our R environment:</p>
<pre><code>#&gt; Loading Reputils
#&gt; Loading required package: BSgenome
#&gt; Loading required package: BiocGenerics
#&gt; Loading required package: parallel
#&gt; 
#&gt; Attaching package: 'BiocGenerics'
#&gt; The following objects are masked from 'package:parallel':
#&gt; 
#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ, clusterExport, clusterMap, parApply, parCapply, parLapply, parLapplyLB, parRapply, parSapply, parSapplyLB
#&gt; The following objects are masked from 'package:Reputils':
#&gt; 
#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ, clusterExport, clusterMap, combine, intersect, parApply, parCapply, parLapply, parLapplyLB, parRapply, parSapply,
#&gt;     parSapplyLB, setdiff, union, which
#&gt; The following objects are masked from 'package:stats':
#&gt; 
#&gt;     IQR, mad, sd, var, xtabs
#&gt; The following objects are masked from 'package:base':
#&gt; 
#&gt;     anyDuplicated, append, as.data.frame, basename, cbind, colMeans, colnames, colSums, dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep, grepl, intersect,
#&gt;     is.unsorted, lapply, lengths, Map, mapply, match, mget, order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce, rowMeans, rownames, rowSums, sapply,
#&gt;     setdiff, sort, table, tapply, union, unique, unsplit, which, which.max, which.min
#&gt; Loading required package: S4Vectors
#&gt; Loading required package: stats4
#&gt; 
#&gt; Attaching package: 'S4Vectors'
#&gt; The following objects are masked from 'package:Reputils':
#&gt; 
#&gt;     first, rename, second, values
#&gt; The following object is masked from 'package:base':
#&gt; 
#&gt;     expand.grid
#&gt; Loading required package: IRanges
#&gt; 
#&gt; Attaching package: 'IRanges'
#&gt; The following objects are masked from 'package:Reputils':
#&gt; 
#&gt;     collapse, desc, shift, slice, trim
#&gt; Loading required package: GenomeInfoDb
#&gt; Loading required package: GenomicRanges
#&gt; Loading required package: Biostrings
#&gt; Loading required package: XVector
#&gt; 
#&gt; Attaching package: 'Biostrings'
#&gt; The following object is masked from 'package:base':
#&gt; 
#&gt;     strsplit
#&gt; Loading required package: rtracklayer
#&gt; ── Attaching packages ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse 1.2.1 ──
#&gt; ✔ ggplot2 3.1.0       ✔ purrr   0.3.2  
#&gt; ✔ tibble  2.1.1       ✔ dplyr   0.8.0.1
#&gt; ✔ tidyr   0.8.3       ✔ stringr 1.4.0  
#&gt; ✔ readr   1.3.1       ✔ forcats 0.4.0
#&gt; ── Conflicts ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
#&gt; ✖ purrr::accumulate() masks Reputils::accumulate()
#&gt; ✖ dplyr::between()    masks Reputils::between()
#&gt; ✖ dplyr::collapse()   masks Biostrings::collapse(), IRanges::collapse(), Reputils::collapse()
#&gt; ✖ dplyr::combine()    masks BiocGenerics::combine(), Reputils::combine()
#&gt; ✖ purrr::compact()    masks XVector::compact()
#&gt; ✖ dplyr::desc()       masks IRanges::desc(), Reputils::desc()
#&gt; ✖ tidyr::expand()     masks S4Vectors::expand(), Reputils::expand()
#&gt; ✖ dplyr::filter()     masks Reputils::filter(), stats::filter()
#&gt; ✖ dplyr::first()      masks S4Vectors::first(), Reputils::first()
#&gt; ✖ dplyr::lag()        masks Reputils::lag(), stats::lag()
#&gt; ✖ dplyr::n()          masks Reputils::n()
#&gt; ✖ dplyr::n_distinct() masks Reputils::n_distinct()
#&gt; ✖ ggplot2::Position() masks BiocGenerics::Position(), Reputils::Position(), base::Position()
#&gt; ✖ purrr::reduce()     masks GenomicRanges::reduce(), IRanges::reduce(), Reputils::reduce()
#&gt; ✖ dplyr::rename()     masks S4Vectors::rename(), Reputils::rename()
#&gt; ✖ dplyr::slice()      masks XVector::slice(), IRanges::slice(), Reputils::slice()
#&gt; ✖ purrr::transpose()  masks Reputils::transpose()
#&gt; ✖ purrr::when()       masks Reputils::when()
#&gt; Curating TE intervals
#&gt; Resolving overlaps
#&gt; Labeling modified intervals</code></pre>
</div>
<div id="introduce-sequence-errors" class="section level2">
<h2 class="hasAnchor">
<a href="#introduce-sequence-errors" class="anchor"></a>Introduce sequence errors</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">bams    &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grep</a></span>(<span class="st">'deduplicated'</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">dir</a></span>(<span class="st">'~/davidbr/proj/epitherapy/data/h1299/10x/dacsb/aligned/chunked_bams/'</span>, <span class="dt">pattern =</span> <span class="st">'.bam$'</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>), <span class="dt">value =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="kw">mutateReads</span>(<span class="dt">BSgenome   =</span> Hsapiens,</a>
<a class="sourceLine" id="cb2-4" data-line-number="4">            <span class="dt">reads      =</span> bams,</a>
<a class="sourceLine" id="cb2-5" data-line-number="5">            <span class="dt">paired     =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb2-6" data-line-number="6">            <span class="dt">tes        =</span> tes,</a>
<a class="sourceLine" id="cb2-7" data-line-number="7">            <span class="dt">n_reads    =</span> <span class="fl">1e6</span>,</a>
<a class="sourceLine" id="cb2-8" data-line-number="8">            <span class="dt">outdir     =</span> <span class="st">'~/davidbr/proj/epitherapy/data/h1299/10x/dacsb/fastqs/'</span>,</a>
<a class="sourceLine" id="cb2-9" data-line-number="9">            <span class="dt">error_rate =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">8</span>))</a></code></pre></div>
</div>
<div id="star-alignment" class="section level2">
<h2 class="hasAnchor">
<a href="#star-alignment" class="anchor"></a>STAR alignment</h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">r1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">dir</a></span>(<span class="st">'~/davidbr/proj/epitherapy/data/h1299/10x/dacsb/fastqs/'</span>, <span class="dt">pattern =</span> <span class="st">'perc_error_R1'</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">r2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">dir</a></span>(<span class="st">'~/davidbr/proj/epitherapy/data/h1299/10x/dacsb/fastqs/'</span>, <span class="dt">pattern =</span> <span class="st">'perc_error_R2'</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">'~/davidbr/proj/epitherapy/data/h1299/10x/dacsb/aligned/'</span>)</a>
<a class="sourceLine" id="cb3-5" data-line-number="5">cmds &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>()</a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(r1))</a>
<a class="sourceLine" id="cb3-7" data-line-number="7">{</a>
<a class="sourceLine" id="cb3-8" data-line-number="8">  cmds[[i]] &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"><span class="st">    </span><span class="kw">glue</span>(<span class="st">"system('/net/mraid14/export/data/tools/star/STAR-2.5.1b/source/STAR --runThreadN 10 --genomeDir /net/mraid14/export/data/users/davidbr/tools/cellranger/references/refdata-cellranger-GRCh38-1.2.0/star/ --outSAMtype BAM SortedByCoordinate --outFilterMultimapScoreRange 2 --outFileNamePrefix {gsub('.fastq', '.', basename(r1[i]))} --outSAMunmapped Within --readFilesIn {r1[i]} {r2[i]}')"</span>)</a>
<a class="sourceLine" id="cb3-10" data-line-number="10">}</a>
<a class="sourceLine" id="cb3-11" data-line-number="11"></a>
<a class="sourceLine" id="cb3-12" data-line-number="12"><span class="co"># create new empty env and fill with relevant</span></a>
<a class="sourceLine" id="cb3-13" data-line-number="13">empty &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/environment">new.env</a></span>(<span class="dt">parent =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/environment">emptyenv</a></span>())</a>
<a class="sourceLine" id="cb3-14" data-line-number="14"></a>
<a class="sourceLine" id="cb3-15" data-line-number="15"><span class="co"># distribute, compute and combine</span></a>
<a class="sourceLine" id="cb3-16" data-line-number="16">res &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb3-17" data-line-number="17"><span class="st">  </span><span class="kw">gcluster.run3</span>(<span class="dt">command_list =</span> cmds,  </a>
<a class="sourceLine" id="cb3-18" data-line-number="18">                  <span class="dt">max.jobs =</span> <span class="dv">50</span>, </a>
<a class="sourceLine" id="cb3-19" data-line-number="19">                  <span class="dt">envir =</span> empty, </a>
<a class="sourceLine" id="cb3-20" data-line-number="20">                  <span class="dt">io_saturation =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
</div>
<div id="import-reads" class="section level2">
<h2 class="hasAnchor">
<a href="#import-reads" class="anchor"></a>Import reads</h2>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">bams  &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grep</a></span>(<span class="st">'error'</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">dir</a></span>(<span class="st">'~/davidbr/proj/epitherapy/data/h1299/10x/dacsb/aligned/'</span>, <span class="dt">pattern =</span> <span class="st">'bam$'</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>), <span class="dt">value =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"></a>
<a class="sourceLine" id="cb4-3" data-line-number="3">reads &lt;-<span class="st"> </span><span class="kw">importBAM</span>(bams,</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">                   <span class="dt">paired  =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb4-5" data-line-number="5">                   <span class="dt">mate    =</span> <span class="st">'first'</span>,</a>
<a class="sourceLine" id="cb4-6" data-line-number="6">                   <span class="dt">anchor  =</span> <span class="st">'fiveprime'</span>,</a>
<a class="sourceLine" id="cb4-7" data-line-number="7">                   <span class="dt">multi   =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb4-8" data-line-number="8">                   <span class="dt">what    =</span> <span class="st">'qname'</span>,</a>
<a class="sourceLine" id="cb4-9" data-line-number="9">                   <span class="dt">barcode =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'err_0.25'</span>, <span class="st">'err_0.5'</span>, <span class="st">'err_0'</span>, <span class="st">'err_1'</span>, <span class="st">'err_2'</span>, <span class="st">'err_4'</span>, <span class="st">'err_8'</span>))</a>
<a class="sourceLine" id="cb4-10" data-line-number="10"><span class="co">#&gt; Importing BAM file(s)</span></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="co">#&gt; Reading /home/davidbr/davidbr/proj/epitherapy/data/h1299/10x/dacsb/aligned//aligned_reads_0.25perc_error_R1.Aligned.sortedByCoord.out.bam</span></a>
<a class="sourceLine" id="cb4-12" data-line-number="12"><span class="co">#&gt; Reading /home/davidbr/davidbr/proj/epitherapy/data/h1299/10x/dacsb/aligned//aligned_reads_0.5perc_error_R1.Aligned.sortedByCoord.out.bam</span></a>
<a class="sourceLine" id="cb4-13" data-line-number="13"><span class="co">#&gt; Reading /home/davidbr/davidbr/proj/epitherapy/data/h1299/10x/dacsb/aligned//aligned_reads_0perc_error_R1.Aligned.sortedByCoord.out.bam</span></a>
<a class="sourceLine" id="cb4-14" data-line-number="14"><span class="co">#&gt; Reading /home/davidbr/davidbr/proj/epitherapy/data/h1299/10x/dacsb/aligned//aligned_reads_1perc_error_R1.Aligned.sortedByCoord.out.bam</span></a>
<a class="sourceLine" id="cb4-15" data-line-number="15"><span class="co">#&gt; Reading /home/davidbr/davidbr/proj/epitherapy/data/h1299/10x/dacsb/aligned//aligned_reads_2perc_error_R1.Aligned.sortedByCoord.out.bam</span></a>
<a class="sourceLine" id="cb4-16" data-line-number="16"><span class="co">#&gt; Reading /home/davidbr/davidbr/proj/epitherapy/data/h1299/10x/dacsb/aligned//aligned_reads_4perc_error_R1.Aligned.sortedByCoord.out.bam</span></a>
<a class="sourceLine" id="cb4-17" data-line-number="17"><span class="co">#&gt; Reading /home/davidbr/davidbr/proj/epitherapy/data/h1299/10x/dacsb/aligned//aligned_reads_8perc_error_R1.Aligned.sortedByCoord.out.bam</span></a>
<a class="sourceLine" id="cb4-18" data-line-number="18"><span class="co">#&gt; Done</span></a></code></pre></div>
</div>
<div id="benchmarking" class="section level2">
<h2 class="hasAnchor">
<a href="#benchmarking" class="anchor"></a>Benchmarking</h2>
<div id="percent-mapped-to-same-te-locus" class="section level3">
<h3 class="hasAnchor">
<a href="#percent-mapped-to-same-te-locus" class="anchor"></a>Percent mapped to same TE locus</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co"># add matching TE locus</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">reads_anno &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="st">  </span>plyranges<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/plyranges/topics/overlap-joins">join_overlap_left_directed</a></span>(reads, tes) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="st">  </span><span class="kw">as_tibble</span>()   </a>
<a class="sourceLine" id="cb5-5" data-line-number="5"></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="co"># exclude reads not mapped or not mapping to TEs at 0% error_rate</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7">reads_anno &lt;-</a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="st">  </span>reads_anno <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-9" data-line-number="9"><span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(qname <span class="op">%in%</span><span class="st"> </span>qname[barcode <span class="op">==</span><span class="st"> 'err_0'</span> <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>(name)])</a>
<a class="sourceLine" id="cb5-10" data-line-number="10">  </a>
<a class="sourceLine" id="cb5-11" data-line-number="11">p_n_reads &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb5-12" data-line-number="12"><span class="st">  </span>reads_anno <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-13" data-line-number="13"><span class="st">  </span><span class="kw">count</span>(barcode) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-14" data-line-number="14"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(barcode, n)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-15" data-line-number="15"><span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat =</span> <span class="st">'identity'</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb5-16" data-line-number="16"><span class="st">    </span><span class="kw">xlab</span>(<span class="st">'Read error (%)'</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb5-17" data-line-number="17"><span class="st">    </span><span class="kw">ylab</span>(<span class="st">'Uniquely mapped reads (#)'</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb5-18" data-line-number="18"><span class="st">    </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle =</span> <span class="dv">90</span>))</a>
<a class="sourceLine" id="cb5-19" data-line-number="19"></a>
<a class="sourceLine" id="cb5-20" data-line-number="20"><span class="co"># get families with min 100 reads</span></a>
<a class="sourceLine" id="cb5-21" data-line-number="21">fam_expr &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb5-22" data-line-number="22"><span class="st">  </span>reads_anno <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-23" data-line-number="23"><span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(barcode <span class="op">==</span><span class="st"> 'err_0'</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-24" data-line-number="24"><span class="st">  </span><span class="kw">count</span>(name) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-25" data-line-number="25"><span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(n <span class="op">&gt;=</span><span class="st"> </span><span class="dv">100</span>)</a>
<a class="sourceLine" id="cb5-26" data-line-number="26"></a>
<a class="sourceLine" id="cb5-27" data-line-number="27">read_stats &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb5-28" data-line-number="28"><span class="st">  </span>reads_anno <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-29" data-line-number="29"><span class="st">    </span><span class="kw">group_by</span>(qname) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-30" data-line-number="30"><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">name0 =</span> name[barcode <span class="op">==</span><span class="st"> 'err_0'</span>],                              <span class="co"># get read TE family mapping under 0% error</span></a>
<a class="sourceLine" id="cb5-31" data-line-number="31">           <span class="dt">same_locus =</span> id_unique <span class="op">==</span><span class="st"> </span>id_unique[barcode <span class="op">==</span><span class="st"> 'err_0'</span>]) <span class="op">%&gt;%</span><span class="st">   </span><span class="co"># check if TE locus mapping is same as under 0% error</span></a>
<a class="sourceLine" id="cb5-32" data-line-number="32"><span class="st">    </span>ungroup <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-33" data-line-number="33"><span class="st">    </span><span class="kw">replace_na</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">same_locus =</span> <span class="ot">FALSE</span>)) <span class="op">%&gt;%</span><span class="st">                              </span></a>
<a class="sourceLine" id="cb5-34" data-line-number="34"><span class="st">    </span><span class="kw">group_by</span>(name0, barcode) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-35" data-line-number="35"><span class="st">    </span><span class="kw">summarize</span>(<span class="dt">perc_same =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>(same_locus) <span class="op">/</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(same_locus) <span class="op">*</span><span class="st"> </span><span class="dv">100</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-36" data-line-number="36"><span class="st">    </span>ungroup</a>
<a class="sourceLine" id="cb5-37" data-line-number="37">    </a>
<a class="sourceLine" id="cb5-38" data-line-number="38"><span class="co"># annotate and filter</span></a>
<a class="sourceLine" id="cb5-39" data-line-number="39">read_stats_f &lt;-</a>
<a class="sourceLine" id="cb5-40" data-line-number="40"><span class="st">  </span>read_stats <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-41" data-line-number="41"><span class="st">    </span><span class="kw">inner_join</span>(.,                            <span class="co"># filter for expressed families</span></a>
<a class="sourceLine" id="cb5-42" data-line-number="42">               fam_expr <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-43" data-line-number="43"><span class="st">                </span><span class="kw">rename</span>(<span class="dt">name0 =</span> name)</a>
<a class="sourceLine" id="cb5-44" data-line-number="44">              ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-45" data-line-number="45"><span class="st">    </span><span class="kw">left_join</span>(.,                             <span class="co"># add repclass annotation</span></a>
<a class="sourceLine" id="cb5-46" data-line-number="46">              reads_anno <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-47" data-line-number="47"><span class="st">                </span><span class="kw">rename</span>(<span class="dt">name0 =</span> name) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-48" data-line-number="48"><span class="st">                </span><span class="kw">select</span>(name0, repclass) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-49" data-line-number="49"><span class="st">                </span>distinct</a>
<a class="sourceLine" id="cb5-50" data-line-number="50">              ) </a>
<a class="sourceLine" id="cb5-51" data-line-number="51"><span class="co">#&gt; Joining, by = "name0"</span></a>
<a class="sourceLine" id="cb5-52" data-line-number="52"><span class="co">#&gt; Joining, by = "name0"</span></a>
<a class="sourceLine" id="cb5-53" data-line-number="53">              </a>
<a class="sourceLine" id="cb5-54" data-line-number="54"><span class="co"># label worst families</span></a>
<a class="sourceLine" id="cb5-55" data-line-number="55">read_stats_f &lt;-</a>
<a class="sourceLine" id="cb5-56" data-line-number="56"><span class="st">  </span>read_stats_f <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-57" data-line-number="57"><span class="st">  </span><span class="kw">group_by</span>(repclass) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-58" data-line-number="58"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">worst =</span> perc_same <span class="op">==</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Extremes">min</a></span>(perc_same)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-59" data-line-number="59"><span class="st">  </span>ungroup <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-60" data-line-number="60"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">label =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(worst, name0, <span class="st">''</span>))              </a>
<a class="sourceLine" id="cb5-61" data-line-number="61"></a>
<a class="sourceLine" id="cb5-62" data-line-number="62"><span class="co"># plot</span></a>
<a class="sourceLine" id="cb5-63" data-line-number="63">p_same_perc &lt;-</a>
<a class="sourceLine" id="cb5-64" data-line-number="64"><span class="st">  </span>read_stats_f <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-65" data-line-number="65"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> barcode, <span class="dt">y =</span> perc_same, <span class="dt">group =</span> name0, <span class="dt">col =</span> repclass, <span class="dt">label =</span> label)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-66" data-line-number="66"><span class="st">      </span><span class="kw">geom_point</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb5-67" data-line-number="67"><span class="st">      </span><span class="kw">geom_line</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb5-68" data-line-number="68"><span class="st">      </span><span class="kw">facet_wrap</span>(<span class="op">~</span>repclass) <span class="op">+</span></a>
<a class="sourceLine" id="cb5-69" data-line-number="69"><span class="st">      </span><span class="kw">scale_color_brewer</span>(<span class="dt">palette =</span> <span class="st">'Set1'</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb5-70" data-line-number="70"><span class="st">      </span>ggrepel<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggrepel/topics/geom_text_repel">geom_label_repel</a></span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb5-71" data-line-number="71"><span class="st">      </span><span class="kw">xlab</span>(<span class="st">'Read error (%)'</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb5-72" data-line-number="72"><span class="st">      </span><span class="kw">ylab</span>(<span class="st">'Reads mapping to true TE locus (%)'</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-73" data-line-number="73"><span class="st">      </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">'none'</span>,</a>
<a class="sourceLine" id="cb5-74" data-line-number="74">            <span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle =</span> <span class="dv">90</span>))</a></code></pre></div>
</div>
<div id="ltr12c-umi-counts" class="section level3">
<h3 class="hasAnchor">
<a href="#ltr12c-umi-counts" class="anchor"></a>LTR12C UMI counts</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">ltr12c &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="st">  </span>reads_anno <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="st">    </span>as_tibble <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(name <span class="op">==</span><span class="st"> 'LTR12C'</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="st">    </span><span class="kw">group_by</span>(id_unique, barcode) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="st">    </span><span class="kw">summarise</span>(<span class="dt">counts =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>(NH_weight)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="st">    </span>ungroup <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-8" data-line-number="8"><span class="st">    </span><span class="kw">spread</span>(barcode, counts, <span class="dt">fill =</span> <span class="dv">0</span>)             </a>
<a class="sourceLine" id="cb6-9" data-line-number="9">                </a>
<a class="sourceLine" id="cb6-10" data-line-number="10"><span class="co"># plot</span></a>
<a class="sourceLine" id="cb6-11" data-line-number="11">p1 &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb6-12" data-line-number="12"><span class="st">  </span>ltr12c <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-13" data-line-number="13"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> err_<span class="dv">0</span>, <span class="dt">y =</span> err_<span class="fl">0.25</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-14" data-line-number="14"><span class="st">      </span>ggrastr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggrastr/topics/geom_point_rast">geom_point_rast</a></span>(<span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-15" data-line-number="15"><span class="st">      </span><span class="kw">scale_y_log10</span>(<span class="dt">limits =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>, <span class="fl">1e4</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-16" data-line-number="16"><span class="st">      </span><span class="kw">scale_x_log10</span>(<span class="dt">limits =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>, <span class="fl">1e4</span>))  </a>
<a class="sourceLine" id="cb6-17" data-line-number="17"></a>
<a class="sourceLine" id="cb6-18" data-line-number="18"><span class="co"># plot</span></a>
<a class="sourceLine" id="cb6-19" data-line-number="19">p2 &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb6-20" data-line-number="20"><span class="st">  </span>ltr12c <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-21" data-line-number="21"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> err_<span class="dv">0</span>, <span class="dt">y =</span> err_<span class="fl">0.5</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-22" data-line-number="22"><span class="st">      </span>ggrastr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggrastr/topics/geom_point_rast">geom_point_rast</a></span>(<span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-23" data-line-number="23"><span class="st">      </span><span class="kw">scale_y_log10</span>(<span class="dt">limits =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>, <span class="fl">1e4</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-24" data-line-number="24"><span class="st">      </span><span class="kw">scale_x_log10</span>(<span class="dt">limits =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>, <span class="fl">1e4</span>))  </a>
<a class="sourceLine" id="cb6-25" data-line-number="25">      </a>
<a class="sourceLine" id="cb6-26" data-line-number="26">p3 &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb6-27" data-line-number="27"><span class="st">  </span>ltr12c <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-28" data-line-number="28"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> err_<span class="dv">0</span>, <span class="dt">y =</span> err_<span class="dv">1</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-29" data-line-number="29"><span class="st">      </span>ggrastr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggrastr/topics/geom_point_rast">geom_point_rast</a></span>(<span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-30" data-line-number="30"><span class="st">      </span><span class="kw">scale_y_log10</span>(<span class="dt">limits =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>, <span class="fl">1e4</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-31" data-line-number="31"><span class="st">      </span><span class="kw">scale_x_log10</span>(<span class="dt">limits =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>, <span class="fl">1e4</span>))     </a>
<a class="sourceLine" id="cb6-32" data-line-number="32"></a>
<a class="sourceLine" id="cb6-33" data-line-number="33">p4 &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb6-34" data-line-number="34"><span class="st">  </span>ltr12c <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-35" data-line-number="35"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> err_<span class="dv">0</span>, <span class="dt">y =</span> err_<span class="dv">2</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-36" data-line-number="36"><span class="st">      </span>ggrastr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggrastr/topics/geom_point_rast">geom_point_rast</a></span>(<span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-37" data-line-number="37"><span class="st">      </span><span class="kw">scale_y_log10</span>(<span class="dt">limits =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>, <span class="fl">1e4</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-38" data-line-number="38"><span class="st">      </span><span class="kw">scale_x_log10</span>(<span class="dt">limits =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>, <span class="fl">1e4</span>))  </a>
<a class="sourceLine" id="cb6-39" data-line-number="39"></a>
<a class="sourceLine" id="cb6-40" data-line-number="40">p5 &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb6-41" data-line-number="41"><span class="st">  </span>ltr12c <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-42" data-line-number="42"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> err_<span class="dv">0</span>, <span class="dt">y =</span> err_<span class="dv">4</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-43" data-line-number="43"><span class="st">      </span>ggrastr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggrastr/topics/geom_point_rast">geom_point_rast</a></span>(<span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-44" data-line-number="44"><span class="st">      </span><span class="kw">scale_y_log10</span>(<span class="dt">limits =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>, <span class="fl">1e4</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-45" data-line-number="45"><span class="st">      </span><span class="kw">scale_x_log10</span>(<span class="dt">limits =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>, <span class="fl">1e4</span>))      </a>
<a class="sourceLine" id="cb6-46" data-line-number="46"></a>
<a class="sourceLine" id="cb6-47" data-line-number="47">p6 &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb6-48" data-line-number="48"><span class="st">  </span>ltr12c <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-49" data-line-number="49"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> err_<span class="dv">0</span>, <span class="dt">y =</span> err_<span class="dv">8</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-50" data-line-number="50"><span class="st">      </span>ggrastr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggrastr/topics/geom_point_rast">geom_point_rast</a></span>(<span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-51" data-line-number="51"><span class="st">      </span><span class="kw">scale_y_log10</span>(<span class="dt">limits =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>, <span class="fl">1e4</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb6-52" data-line-number="52"><span class="st">      </span><span class="kw">scale_x_log10</span>(<span class="dt">limits =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>, <span class="fl">1e4</span>))       </a>
<a class="sourceLine" id="cb6-53" data-line-number="53">      </a>
<a class="sourceLine" id="cb6-54" data-line-number="54">p_comb1 &lt;-<span class="st"> </span>cowplot<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/cowplot/topics/plot_grid">plot_grid</a></span>(p_n_reads, p_same_perc, <span class="dt">rel_widths =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>))</a>
<a class="sourceLine" id="cb6-55" data-line-number="55">p_comb2 &lt;-<span class="st"> </span>cowplot<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/cowplot/topics/plot_grid">plot_grid</a></span>(p1, p2, p3, p4, p5, p6)</a></code></pre></div>
</div>
<div id="plotting" class="section level3">
<h3 class="hasAnchor">
<a href="#plotting" class="anchor"></a>Plotting</h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">cowplot<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/cowplot/topics/plot_grid">plot_grid</a></span>(p_comb1, p_comb2, <span class="dt">ncol =</span> <span class="dv">1</span>)</a></code></pre></div>
<p><img src="Mapping_QC_files/figure-html/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#benchmark-read-alignments-to-transposable-elements">Benchmark read alignments to transposable elements</a><ul class="nav nav-pills nav-stacked">
<li><a href="#getting-started">Getting started</a></li>
      <li><a href="#introduce-sequence-errors">Introduce sequence errors</a></li>
      <li><a href="#star-alignment">STAR alignment</a></li>
      <li><a href="#import-reads">Import reads</a></li>
      <li><a href="#benchmarking">Benchmarking</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://scholar.google.de/citations?user=ossdDj4AAAAJ&amp;hl=en&amp;oi=ao">David Brocks</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
